# 2. Generando Incrustaciones contextuales (Contextual embeddings)

## 2.1 Carga y transformacion de informacion

In [21]:
#!pip install openpyxl
#!pip install demoji
#!pip install unidecode
#!pip install sentence-transformers umap-learn plotly scikit-learn tensorflow

In [15]:
from pathlib import Path
import numpy as np
import pandas as pd
import re
from unidecode import unidecode
import seaborn as sns
import matplotlib.pyplot as plt

# ============================================================================
# Rutas del proyecto
# ============================================================================
PROJECT_DIR = Path(r"D:\Datasets\atribucion-autoria-tweets")
DATA_DIR = PROJECT_DIR / "data"
PROCESSED_DIR = DATA_DIR / "processed"

# Archivo de entrada
data_path = PROCESSED_DIR / "fullclases.csv"

# ============================================================================
# Carga de datos
# ============================================================================
merged_data = pd.read_csv(data_path)
merged_data.info(verbose=True, memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 963498 entries, 0 to 963497
Data columns (total 17 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   id             963498 non-null  int64 
 1   tweetText      963498 non-null  object
 2   tweetURL       963498 non-null  object
 3   type           963498 non-null  object
 4   tweetAuthor    959858 non-null  object
 5   handle         963498 non-null  object
 6   replyCount     963498 non-null  int64 
 7   quoteCount     963498 non-null  int64 
 8   retweetCount   963498 non-null  int64 
 9   likeCount      963498 non-null  int64 
 10  views          963498 non-null  object
 11  bookmarkCount  963498 non-null  int64 
 12  createdAt      963498 non-null  object
 13  allMediaURL    250166 non-null  object
 14  videoURL       65276 non-null   object
 15  filename       963498 non-null  object
 16  Final_user     963498 non-null  object
dtypes: int64(6), object(11)
memory usage: 912.8 MB


In [16]:
merged_data.head()

,id,tweetText,tweetURL,type,tweetAuthor,handle,replyCount,quoteCount,retweetCount,likeCount,views,bookmarkCount,createdAt,allMediaURL,videoURL,filename,Final_user
0,1894501879817584955,Pues resulta que el ya famoso despacho morenis...,https://x.com/AccionNacional/status/1894501879...,retweet,Elías Lixa,@eLiasLixa,0,0,2391,0,-,0,2025-02-25 15:37:00,NaN,NaN,TwExtract-AccionNacional-20250226_082459,AccionNacional
1,1894501607640842641,¡Las mujeres y niñas merecen vivir libres de v...,https://x.com/AccionNacional/status/1894501607...,tweet,Acción Nacional,@AccionNacional,3,0,3,3,654,1,2025-02-25 15:35:56,https://pbs.twimg.com/media/GkqfK5VXoAEeuCD.jpg,NaN,TwExtract-AccionNacional-20250226_082459,AccionNacional
2,1894476017907241089,Sostuvimos una productiva reunión de trabajo e...,https://x.com/AccionNacional/status/1894476017...,retweet,Senadores del PAN,@SenadoresdelPAN,0,0,8,0,-,0,2025-02-25 13:54:14,NaN,NaN,TwExtract-AccionNacional-20250226_082459,AccionNacional
3,1894459390746661291,ENTREGA DE LA OBRA DE AMPLIACIÓN DEL PUENTE DE...,https://x.com/AccionNacional/status/1894459390...,retweet,Tere Jiménez,@TereJimenezE,0,0,5,0,-,0,2025-02-25 12:48:10,NaN,NaN,TwExtract-AccionNacional-20250226_082459,AccionNacional
4,1894456464963768747,"¡Muy feliz cumpleaños Senador @RicardoAnayaC, ...",https://x.com/AccionNacional/status/1894456464...,tweet,Acción Nacional,@AccionNacional,117,7,75,426,6800,3,2025-02-25 12:36:33,https://pbs.twimg.com/media/Gkp2HLFW8AAwz5K.jpg,NaN,TwExtract-AccionNacional-20250226_082459,AccionNacional


In [17]:
# List of users to remove
users_to_remove = ['DanielNoboaOk','DiegoFC','jorgeramosnews','JorgeGarcesMx', 'RobertoMadrazo_','LuisitoComunica','RicardoAnayaC', 'AlfredoJalife','PagesBeatriz', 'EstefaniaVeloz']  # Add your users here

# Remove the specified users using query()
merged_data = merged_data.query("Final_user not in @users_to_remove")
merged_data.info(verbose=True, memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Index: 928759 entries, 0 to 963497
Data columns (total 17 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   id             928759 non-null  int64 
 1   tweetText      928759 non-null  object
 2   tweetURL       928759 non-null  object
 3   type           928759 non-null  object
 4   tweetAuthor    928759 non-null  object
 5   handle         928759 non-null  object
 6   replyCount     928759 non-null  int64 
 7   quoteCount     928759 non-null  int64 
 8   retweetCount   928759 non-null  int64 
 9   likeCount      928759 non-null  int64 
 10  views          928759 non-null  object
 11  bookmarkCount  928759 non-null  int64 
 12  createdAt      928759 non-null  object
 13  allMediaURL    239212 non-null  object
 14  videoURL       63109 non-null   object
 15  filename       928759 non-null  object
 16  Final_user     928759 non-null  object
dtypes: int64(6), object(11)
memory usage: 887.4 MB


In [18]:
# Extract unique names from the 'Nombre_Personaje' column
unique_names = merged_data['Final_user'].unique()

# Create the dictionary mapping names to indices starting from 1 to 37
diccionario_nombres = dict(zip(unique_names, range(0, 60)))  # Changed range to (1, 38)
diccionario_nombres

{'AccionNacional': 0,
 'AgustinLaje': 1,
 'alitomorenoc': 2,
 'beltrandelrio': 3,
 'brozoxmiswebs': 4,
 'carlaescoffie': 5,
 'carlosalazraki': 6,
 'cesarpinedar': 7,
 'CFKArgentina': 8,
 'ChumelTorres': 9,
 'ClaudioXGG': 10,
 'DeniseDresserG': 11,
 'DiazCanelB': 12,
 'EdySmol': 13,
 'EVerastegui': 14,
 'fede_bonasso': 15,
 'FelipeCalderon': 16,
 'fernandeznorona': 17,
 'GabrielBoric': 18,
 'GlodeJo07': 19,
 'GloriaAlvarez85': 20,
 'HernanGomezB': 21,
 'jet1403': 22,
 'Jjlopez_almejo': 23,
 'JLMNoticias': 24,
 'JMilei': 25,
 'LAURAZAPATAM': 26,
 'LeonKrauze': 27,
 'LillyTellez': 28,
 'lopezobrador_': 29,
 'MariaCorinaYA': 30,
 'MarkoCortes': 31,
 'MovCiudadanoMX': 32,
 'nayibbukele': 33,
 'NicolasMaduro': 34,
 'PabloIglesias': 35,
 'PartidoMorenaMx': 36,
 'partidoverdemex': 37,
 'petrogustavo': 38,
 'PonchoGutz': 39,
 'RevistaMigala': 40,
 'RicardoBSalinas': 41,
 'sanchezcastejon': 42,
 'VicenteFoxQue': 43,
 'Viri_Rios': 44,
 'XochitlGalvez': 45,
 'YosoyPedrero': 46}

In [19]:
# perform mapping
merged_data['Tweetero_num'] = merged_data['Final_user'].map(diccionario_nombres)

# Keep only needed columns
columns_to_keep = ['tweetText', 'Tweetero_num']
df_filtered = merged_data.drop(columns=[col for col in merged_data.columns if col not in columns_to_keep])

In [20]:
# remove NaN
df_cleaned = df_filtered.dropna(subset=['tweetText'])

In [21]:
df_cleaned

,tweetText,Tweetero_num
0,Pues resulta que el ya famoso despacho morenis...,0
1,¡Las mujeres y niñas merecen vivir libres de v...,0
2,Sostuvimos una productiva reunión de trabajo e...,0
3,ENTREGA DE LA OBRA DE AMPLIACIÓN DEL PUENTE DE...,0
4,"¡Muy feliz cumpleaños Senador @RicardoAnayaC, ...",0
...,...,...
963493,"@Miriam_Junne Hasta aquí contigo, bye",46
963494,"Antes, @lopezobrador_ decía que si de que serv...",46
963495,Spoiler: El único que acabará con el machismo ...,46
963496,@KarlaRiveraMX Si,46


In [23]:
df_cleaned.to_csv(r'D:\Datasets\atribucion-autoria-tweets\data\processed\df_cleaned.csv', index=False)

## 2.2 Implementacion de Modelos

### 2.2.1. SentenceTransformer

In [12]:
import torch
from sentence_transformers import SentenceTransformer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.metrics import classification_report, confusion_matrix
import umap
import plotly.express as px
import plotly.graph_objects as go
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, BatchNormalization, Activation
from tensorflow.keras.regularizers import l2
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau
from tensorflow.keras.optimizers import Adam
import matplotlib.pyplot as plt
import seaborn as sns

print(f"Total de tweets: {len(df_cleaned)}")
print(f"Número de autores únicos: {df_cleaned['Tweetero_num'].nunique()}")

Total de tweets: 928759
Número de autores únicos: 47


In [14]:
# ============================================================================
# 3. GENERACIÓN DE EMBEDDINGS CON SENTENCE TRANSFORMER
# ============================================================================

print("Generando embeddings...")

model = SentenceTransformer('dccuchile/bert-base-spanish-wwm-uncased')

batch_size_textos = 5000       # número de tweets por bloque
batch_size_modelo = 64         # batch interno del modelo

all_embeddings = []

for i in range(0, len(df_cleaned), batch_size_textos):

    print(f"\nProcesando tweets {i:,} - {min(i+batch_size_textos, len(df_cleaned)):,}")

    textos = df_cleaned["tweetText"].iloc[i:i+batch_size_textos].tolist()

    emb = model.encode(
        textos,
        batch_size=batch_size_modelo,
        show_progress_bar=True,
        convert_to_numpy=True,
        normalize_embeddings=True
    )

    all_embeddings.append(emb)

embeddings = np.vstack(all_embeddings)

print(f"\nShape de embeddings: {embeddings.shape}")
print(f"Varianza total: {np.var(embeddings):.6f}")

Generando embeddings...


Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

[transformers] BertModel LOAD REPORT from: dccuchile/bert-base-spanish-wwm-uncased
Key                                        | Status     | 
-------------------------------------------+------------+-
cls.predictions.decoder.weight             | UNEXPECTED | 
cls.predictions.bias                       | UNEXPECTED | 
cls.predictions.transform.dense.bias       | UNEXPECTED | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED | 
cls.predictions.transform.dense.weight     | UNEXPECTED | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED | 
cls.predictions.decoder.bias               | UNEXPECTED | 
pooler.dense.bias                          | MISSING    | 
pooler.dense.weight                        | MISSING    | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.



Procesando tweets 0 - 5,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 5,000 - 10,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 10,000 - 15,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 15,000 - 20,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 20,000 - 25,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 25,000 - 30,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 30,000 - 35,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 35,000 - 40,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 40,000 - 45,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 45,000 - 50,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 50,000 - 55,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 55,000 - 60,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 60,000 - 65,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 65,000 - 70,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 70,000 - 75,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 75,000 - 80,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 80,000 - 85,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 85,000 - 90,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 90,000 - 95,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 95,000 - 100,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 100,000 - 105,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 105,000 - 110,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 110,000 - 115,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 115,000 - 120,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 120,000 - 125,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 125,000 - 130,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 130,000 - 135,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 135,000 - 140,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 140,000 - 145,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 145,000 - 150,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 150,000 - 155,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 155,000 - 160,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 160,000 - 165,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 165,000 - 170,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 170,000 - 175,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 175,000 - 180,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 180,000 - 185,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 185,000 - 190,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 190,000 - 195,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 195,000 - 200,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 200,000 - 205,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 205,000 - 210,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 210,000 - 215,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 215,000 - 220,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 220,000 - 225,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 225,000 - 230,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 230,000 - 235,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 235,000 - 240,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 240,000 - 245,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 245,000 - 250,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 250,000 - 255,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 255,000 - 260,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 260,000 - 265,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 265,000 - 270,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 270,000 - 275,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 275,000 - 280,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 280,000 - 285,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 285,000 - 290,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 290,000 - 295,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 295,000 - 300,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 300,000 - 305,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 305,000 - 310,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 310,000 - 315,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 315,000 - 320,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 320,000 - 325,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 325,000 - 330,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 330,000 - 335,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 335,000 - 340,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 340,000 - 345,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 345,000 - 350,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 350,000 - 355,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 355,000 - 360,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 360,000 - 365,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 365,000 - 370,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 370,000 - 375,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 375,000 - 380,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 380,000 - 385,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 385,000 - 390,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 390,000 - 395,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 395,000 - 400,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 400,000 - 405,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 405,000 - 410,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 410,000 - 415,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 415,000 - 420,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 420,000 - 425,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 425,000 - 430,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 430,000 - 435,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 435,000 - 440,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 440,000 - 445,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 445,000 - 450,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 450,000 - 455,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 455,000 - 460,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 460,000 - 465,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 465,000 - 470,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 470,000 - 475,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 475,000 - 480,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 480,000 - 485,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 485,000 - 490,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 490,000 - 495,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 495,000 - 500,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 500,000 - 505,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 505,000 - 510,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 510,000 - 515,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 515,000 - 520,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 520,000 - 525,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 525,000 - 530,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 530,000 - 535,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 535,000 - 540,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 540,000 - 545,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 545,000 - 550,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 550,000 - 555,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 555,000 - 560,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 560,000 - 565,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 565,000 - 570,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 570,000 - 575,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 575,000 - 580,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 580,000 - 585,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 585,000 - 590,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 590,000 - 595,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 595,000 - 600,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 600,000 - 605,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 605,000 - 610,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 610,000 - 615,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 615,000 - 620,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 620,000 - 625,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 625,000 - 630,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 630,000 - 635,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 635,000 - 640,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 640,000 - 645,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 645,000 - 650,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 650,000 - 655,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 655,000 - 660,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 660,000 - 665,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 665,000 - 670,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 670,000 - 675,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 675,000 - 680,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 680,000 - 685,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 685,000 - 690,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 690,000 - 695,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 695,000 - 700,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 700,000 - 705,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 705,000 - 710,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 710,000 - 715,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 715,000 - 720,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 720,000 - 725,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 725,000 - 730,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 730,000 - 735,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 735,000 - 740,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 740,000 - 745,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 745,000 - 750,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 750,000 - 755,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 755,000 - 760,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 760,000 - 765,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 765,000 - 770,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 770,000 - 775,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 775,000 - 780,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 780,000 - 785,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 785,000 - 790,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 790,000 - 795,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 795,000 - 800,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 800,000 - 805,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 805,000 - 810,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 810,000 - 815,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 815,000 - 820,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 820,000 - 825,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 825,000 - 830,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 830,000 - 835,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 835,000 - 840,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 840,000 - 845,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 845,000 - 850,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 850,000 - 855,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 855,000 - 860,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 860,000 - 865,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 865,000 - 870,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 870,000 - 875,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 875,000 - 880,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 880,000 - 885,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 885,000 - 890,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 890,000 - 895,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 895,000 - 900,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 900,000 - 905,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 905,000 - 910,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 910,000 - 915,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 915,000 - 920,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 920,000 - 925,000


Batches:   0%|          | 0/79 [00:00<?, ?it/s]


Procesando tweets 925,000 - 928,759


Batches:   0%|          | 0/59 [00:00<?, ?it/s]


Shape de embeddings: (928759, 768)
Varianza total: 0.001297


In [23]:
# Etiquetas numéricas de los autores
y_encoded = df_cleaned["Tweetero_num"].to_numpy()

print(f"Embeddings: {embeddings.shape}")
print(f"Etiquetas: {y_encoded.shape}")
print(f"Número de clases: {len(np.unique(y_encoded))}")

Embeddings: (928759, 768)
Etiquetas: (928759,)
Número de clases: 47


In [24]:
from pathlib import Path

output_dir = Path(r"D:\Datasets\atribucion-autoria-tweets\data\embeddings\SentenceTransformer")

np.save(output_dir / "embeddings_sentence_transformer.npy", embeddings)
np.save(output_dir / "y_autores.npy", y_encoded)

print("Archivos guardados.")

Archivos guardados.


In [25]:
import joblib

joblib.dump(
    diccionario_nombres,
    output_dir / "diccionario_autores.pkl"
)

print("✅ Diccionario guardado")

✅ Diccionario guardado


In [26]:
emb = np.load(output_dir / "embeddings_sentence_transformer.npy")
y = np.load(output_dir / "y_autores.npy")

print("Embeddings:", emb.shape)
print("Etiquetas:", y.shape)

assert emb.shape == embeddings.shape
assert y.shape == y_encoded.shape

print("✅ Verificación exitosa")

Embeddings: (928759, 768)
Etiquetas: (928759,)
✅ Verificación exitosa


### 2.2.2. Mex_state

Este modelo basado en Roberta se entrenó usando más de 140 millones de tweets de México en español. Recolectados entre diciembre del 2015 y febrero del 2023. 

[Link](https://huggingface.co/guillermoruiz/mex_state)

#### 2.2.2.1. Usando batches y generando varios archivos

In [16]:
from transformers import AutoTokenizer, AutoModelForMaskedLM
import torch
import numpy as np
import os

# ============================================================================
# 1. CARGAR MODELO
# ============================================================================

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Dispositivo: {device}")

tokenizer = AutoTokenizer.from_pretrained("guillermoruiz/mex_state")

model = AutoModelForMaskedLM.from_pretrained(
    "guillermoruiz/mex_state",
    output_hidden_states=True
)

model.to(device)
model.eval()

Dispositivo: cpu


Loading weights:   0%|          | 0/106 [00:00<?, ?it/s]

RobertaForMaskedLM(
  (roberta): RobertaModel(
    (embeddings): RobertaEmbeddings(
      (word_embeddings): Embedding(30006, 512, padding_idx=1)
      (token_type_embeddings): Embedding(2, 512)
      (LayerNorm): LayerNorm((512,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
      (position_embeddings): Embedding(512, 512, padding_idx=1)
    )
    (encoder): RobertaEncoder(
      (layer): ModuleList(
        (0-5): 6 x RobertaLayer(
          (attention): RobertaAttention(
            (self): RobertaSelfAttention(
              (query): Linear(in_features=512, out_features=512, bias=True)
              (key): Linear(in_features=512, out_features=512, bias=True)
              (value): Linear(in_features=512, out_features=512, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): RobertaSelfOutput(
              (dense): Linear(in_features=512, out_features=512, bias=True)
              (LayerNorm): La

In [17]:
# ============================================================================
# 2. RUTA DE SALIDA
# ============================================================================
output_dir_mx = r"D:\Datasets\atribucion-autoria-tweets\data\embeddings\mex_state"
os.makedirs(output_dir_mx, exist_ok=True)

In [18]:
# ============================================================================
# 3. FUNCIÓN PARA EMBEDDINGS
# ============================================================================
def get_embeddings(texts, batch_size=64):
    all_embeddings = []

    for i in range(0, len(texts), batch_size):

        batch = ["[MASK] _GEO " + str(t) for t in texts[i:i+batch_size]]

        tokens = tokenizer(
            batch,
            return_tensors="pt",
            max_length=256,
            padding="max_length",
            truncation=True
        )

        tokens = {
            "input_ids": tokens["input_ids"].to(device),
            "attention_mask": tokens["attention_mask"].to(device)
        }

        with torch.no_grad():
            outputs = model(**tokens)

        emb = outputs.hidden_states[-1][:, 1]

        all_embeddings.append(emb.cpu())

    return torch.cat(all_embeddings)

In [19]:
chunk_size = 5000

In [20]:
# ============================================================================
# 4. GENERAR EMBEDDINGS
# ============================================================================

texts = df_cleaned["tweetText"].astype(str).tolist()

chunk_size = 5000

for start in range(0, len(texts), chunk_size):

    end = min(start + chunk_size, len(texts))

    print(f"\nProcesando {start:,} - {end:,}")

    emb = get_embeddings(
        texts[start:end],
        batch_size=64
    )

    filename = os.path.join(
        output_dir_mx,
        f"embeddings_{start}_{end}.pt"
    )

    torch.save(emb, filename)

    print(f"Guardado: {filename}")

    del emb

    if torch.cuda.is_available():
        torch.cuda.empty_cache()


Procesando 0 - 5,000
Guardado: D:\Datasets\atribucion-autoria-tweets\data\embeddings\mex_state\embeddings_0_5000.pt

Procesando 5,000 - 10,000
Guardado: D:\Datasets\atribucion-autoria-tweets\data\embeddings\mex_state\embeddings_5000_10000.pt

Procesando 10,000 - 15,000
Guardado: D:\Datasets\atribucion-autoria-tweets\data\embeddings\mex_state\embeddings_10000_15000.pt

Procesando 15,000 - 20,000
Guardado: D:\Datasets\atribucion-autoria-tweets\data\embeddings\mex_state\embeddings_15000_20000.pt

Procesando 20,000 - 25,000
Guardado: D:\Datasets\atribucion-autoria-tweets\data\embeddings\mex_state\embeddings_20000_25000.pt

Procesando 25,000 - 30,000
Guardado: D:\Datasets\atribucion-autoria-tweets\data\embeddings\mex_state\embeddings_25000_30000.pt

Procesando 30,000 - 35,000
Guardado: D:\Datasets\atribucion-autoria-tweets\data\embeddings\mex_state\embeddings_30000_35000.pt

Procesando 35,000 - 40,000
Guardado: D:\Datasets\atribucion-autoria-tweets\data\embeddings\mex_state\embeddings_3500

#### 2.2.2.2. Batches en un solo archivo

In [8]:
import torch
import gc
from transformers import AutoTokenizer, AutoModelForMaskedLM
from tqdm import tqdm

# Configuración
DEVICE = "cuda:0" if torch.cuda.is_available() else "cpu"
BATCH_SIZE = 8  # Reduce según tu memoria RAM
CHUNK_SIZE = 10000  # Tamaño del chunk para guardar

print(f"Usando dispositivo: {DEVICE}")

# Cargar modelo y tokenizer
tokenizer = AutoTokenizer.from_pretrained("guillermoruiz/mex_state")
model = AutoModelForMaskedLM.from_pretrained(
    "guillermoruiz/mex_state", 
    output_hidden_states=True
)
model = model.to(DEVICE)
model.eval()

def process_tweets(text_list, bs=BATCH_SIZE, chunk_size=CHUNK_SIZE):
    """Procesa tweets en chunks con checkpointing"""
    
    total_chunks = (len(text_list) + chunk_size - 1) // chunk_size
    all_embeddings = []
    
    for chunk_idx in tqdm(range(total_chunks), desc="Procesando chunks"):
        start = chunk_idx * chunk_size
        end = min((chunk_idx + 1) * chunk_size, len(text_list))
        chunk_text = text_list[start:end]
        
        # Preparar textos para chunk
        chunk_text = ["[MASK] _GEO " + t for t in chunk_text]
        chunk_embeddings = []
        
        for i in range(0, len(chunk_text), bs):
            # Tokenizar
            tokens = tokenizer(
                chunk_text[i:i+bs],
                return_tensors="pt",
                max_length=256,
                padding="max_length",
                truncation=True
            )
            
            # Mover a dispositivo
            input_ids = tokens['input_ids'].to(DEVICE)
            attention_mask = tokens['attention_mask'].to(DEVICE)
            
            # Inferencia
            with torch.no_grad():
                outputs = model(input_ids=input_ids, attention_mask=attention_mask)
            
            # Extraer embeddings
            embeddings = outputs.hidden_states[-1][:, 1].cpu()
            chunk_embeddings.append(embeddings)
            
            # Limpiar memoria
            del tokens, input_ids, attention_mask, outputs, embeddings
            torch.cuda.empty_cache()
        
        # Concatenar embeddings del chunk
        chunk_embeddings = torch.cat(chunk_embeddings)
        
        # Guardar chunk en disco
        chunk_file = f"embeddings_chunk_{chunk_idx}.pt"
        torch.save(chunk_embeddings, chunk_file)
        all_embeddings.append(chunk_file)
        
        # Limpiar memoria
        del chunk_embeddings
        gc.collect()
    
    return all_embeddings

# Ejecutar
text = df_cleaned['tweetText'].astype(str).tolist()
embedding_files = process_tweets(text)

print(f"Procesados {len(text)} tweets en {len(embedding_files)} archivos")

# Para cargar todos los embeddings después:
all_embeddings = []
for file in embedding_files:
    all_embeddings.append(torch.load(file))
final_embeddings = torch.cat(all_embeddings)

Usando dispositivo: cpu


Loading weights:   0%|          | 0/106 [00:00<?, ?it/s]

Procesando chunks: 100%|██████████████████████████████████████████████████████████| 93/93 [27:23:06<00:00, 1060.08s/it]


Procesados 928759 tweets en 93 archivos


##### Exportar objeto pytorch con los embeddings

In [10]:
torch.save(final_embeddings, 'D:/Datasets/atribucion-autoria-tweets/data/embeddings/mex_state_local/embeddings_47c_7000.pt') 

### 2.2.3. RoBERTuito

RoBERTuito es un modelo de lenguaje preentrenado para contenido generado por usuarios en español, entrenado siguiendo las directrices de RoBERTa con 500 millones de tweets. RoBERTuito está disponible en 3 versiones: con mayúsculas y minúsculas, sin mayúsculas y minúsculas, y sin acentos.

[Link](https://github.com/sullivansantana/robertuito/tree/main)

**Nota: Los embeddings de RoBERTuito se generaron externamente debido al tiempo de cómputo requerido. Se utilizan en el Notebook 03 y 04 para clasificación con red neuronal y las maquinas de Soporte vectorial.**